### PR TITLE
machines: Bridge forward mode to support macvtap

### PR DESCRIPTION
**Pull request by skobyda**
This introduces another forward option for Virtual Network Creation: "bridge".
For this option IP configuration is not allowed. User however can choose one or multiple network devices to "be used as direct connection via macvtap" (https://libvirt.org/formatnetwork.html)

Solves: https://github.com/cockpit-project/cockpit/issues/13558

Screenshot from 2020-02-28 15-25-45

### CODE REVIEW

**Comment by skobyda** [comment]
PF4 would need some customizing, but id could be done. I'm also thinking if PF4 simple list would not be better for our case: https://www.patternfly.org/v4/documentation/react/components/simplelist

**Comment by garrett** [comment]
No, it should be a list of checkboxes, not an un-obvious toggle list.

(More than one can be selected, right? If so → list of checkboxes. If only one → list of radio widgets.)

**Comment by skobyda** [comment]
Alright, I updated design to use PF4 tables:
Screenshot from 2020-03-10 13-45-32

Also I had to introduce a field "Interface Devices" into the network overview, because there was no way user could see what interface devices they just assigned during network creation. Link to various network devices are clickable and will redirect you to networks device's page
Screenshot from 2020-03-10 13-45-10

**Comment by garrett** [comment]
PF4 breakpoints are completely bogus with this...
Screenshot_2020-03-10 Virtual Machines - Rain

Therefore, I'm not sure if this is the right set of widgets to use for a list like this.

**Comment by skobyda** [comment]
@garrett This is weird. How can I reproduce this situation? Is this browser-specific?

**Comment by garrett** [comment]
Resize your window. Make it less wide.

This is an issue we're fighting with PF4 at the moment, for all lists. It's especially problematic because sizing is different due to iframes... but using table/list widgets in a dialog would probably continue to be problematic.

**Comment by skobyda** [comment]
@garrett Our cockpit-listing component doesn't have these issues. So perhaps it would be worth it to update that component to have selectable items?

**Comment by garrett** [comment]
Yep, it doesn't... but it's using PF3 + custom code at the moment... that's why.

**Comment by garrett** [comment]
(It's also not as accessible as it should be.)

**Comment by garrett** [comment]
It would be nice to basically have something like this:

https://www.patternfly.org/v4/documentation/core/components/select#checkbox-expanded

But without the dropdown part... just the content area with the checkboxes, and to have it scroll when needed.

I'm not sure how to achieve something like that in React. :disappointed:

**Comment by skobyda** [comment]
@garrett  And what about different approach using input field. 
https://www.patternfly.org/v4/documentation/react/components/select#multiple-with-custom-objects

**Comment by skobyda** [comment]
Also cockpit-listing doesn't use any PF3. It doesn't use any PF at all. It's all custom code. But the point about not being accessible enough stands.

**Comment by garrett** [comment]
No, it shouldn't really be a dropdown. All we want is a simple list with checkboxes.

I guess we could do something like an element with a `ct-checklist` class which would wrap the PF4 checkboxes (or even radio boxes) which would then style them up like the dropdown, with proper spacing, hover effects, and an overflow for a scrollbar (when needed).

We could then make it a widget if we wanted (or just resort to using a div to wrap the group with the class... it's simple enough).

**Comment by garrett** [comment]
I've been playing around with some HTML in my sandbox with PF4 and have found the following HTML code provides basically what we want:

```html
<fieldset class="pf-c-select">

    <label class="pf-c-check pf-c-select__menu-item" for="select-checkbox-expanded-active">
        <input class="pf-c-check__input" type="checkbox" id="select-checkbox-expanded-active" name="select-checkbox-expanded-active">
        <span class="pf-c-check__label">Active</span>
    </label>

    <label class="pf-c-check pf-c-select__menu-item" for="select-checkbox-expanded-cancelled">
        <input class="pf-c-check__input" type="checkbox" id="select-checkbox-expanded-cancelled" name="select-checkbox-expanded-cancelled" checked="">
        <span class="pf-c-check__label">Cancelled</span>
    </label>

    <label class="pf-c-check pf-c-select__menu-item" for="select-checkbox-expanded-paused">
        <input class="pf-c-check__input" type="checkbox" id="select-checkbox-expanded-paused" name="select-checkbox-expanded-paused" checked="">
        <span class="pf-c-check__label">Paused</span>
    </label>

    <label class="pf-c-check pf-c-select__menu-item" for="select-checkbox-expanded-warning">
        <input class="pf-c-check__input" type="checkbox" id="select-checkbox-expanded-warning" name="select-checkbox-expanded-warning">
        <span class="pf-c-check__label">Warning</span>
    </label>

    <label class="pf-c-check pf-c-select__menu-item" for="select-checkbox-expanded-restarted">
        <input class="pf-c-check__input" type="checkbox" id="select-checkbox-expanded-restarted" name="select-checkbox-expanded-restarted" checked="">
        <span class="pf-c-check__label">Restarted</span>
    </label>

</fieldset>
```

It's based on  https://www.patternfly.org/v4/documentation/core/components/select#checkbox-expanded

So, what this means is you should be able to do something like:

1. Use `<fieldset className="pf-c-select">` as a wrapper.
2. Use checkbox inputs as usual in React (provided they result in being wrapped with the label).
3. Give the inputs a `pf-c-select__menu-item` class at the label level (which I guess is just adding a class to the widget... probably?)

And that's it. It acts like the widget we want and even has an overflow baked-in. No custom Cockpit wrapper, no special CSS... unless we want one, of course.

**Comment by skobyda** [comment]
Thanks, that approach works. Here's the freshest update:

Screenshot from 2020-03-17 13-06-40

**Comment by skobyda** [comment]
I have moved the listing here into general component for cockpit-components-listing: **SelectableListing**: Listing with clickable rows to be used in dialogs. I will futher work to use this listing also for bootorder and possibly other places where we user clickable listing in dialogs (like networking page)

File: pkg/lib/cockpit-components-listing.jsx [hunk 1]
```diff
@@ -403,3 +405,35 @@ Listing.propTypes = {
...
+export const SelectableListing = ({ rows, onRowToggle, idPrefix }) => {
+    return (
+        <fieldset className="list-group dialog-list-ct pf-c-select" id={idPrefix + "-selectable-listing"}>
+            {rows.map(row => {
+                return (
+                    <label className="pf-c-check pf-c-select__menu-item" htmlFor={idPrefix + "-" + row.name} key={row.name}>
+                        <input className="pf-c-check__input"
+                            name="select-checkbox-expanded-active"
+                            onChange={() => onRowToggle(row.name)}
+                            type="checkbox"
```
[the remark this answers is not in the document]
Comment (skobyda) [reply pkg/lib/cockpit-components-listing.jsx]: Good comment. I'm working on bigger PR where I want to add additional functionality for this so it can be used for bootorder VM configuration. I will add it there.

**Comment by skobyda** [comment]
Fixed

**Comment by skobyda** [comment]
@KKoukiou could you review?

**Comment by garrett** [comment]
> Also I had to introduce a field "Interface Devices" into the network overview, because there was no way user could see what interface devices they just assigned during network creation. Link to various network devices are clickable and will redirect you to networks device's page
> Screenshot from 2020-03-10 13-45-32

FWIW, the tab spacing is weird here, and the delete button should be on the inside, I think. As it is right now, on Cockpit, we have the default action on the edge (in this case, the right side), so it should be Delete on the left and Activate/Deactivate on the right.

However, we're planning to change it (in at least some places) soon to have the default on the row itself with a … menu for other stuff like delete. But that's near-future stuff, not current stuff. And I don't want it to get in the way of this PR.

**Comment by marusak** [comment]
This one seems like almost done? Now it has bunch of conflicts though. Can you please rebase and then re-ask for review? Thanks!

**Comment by martinpitt** [comment]
Converting to draft to clean up the "needs review" list. Please fix conflicts and the tests, then click "Ready for review". Thanks!

**Comment by skobyda** [comment]
time to reopen this. I rebased and the design looks still same (the new thing is selecting multiple bridges):
Screenshot from 2021-02-15 14-33-48

**Comment by skobyda** [comment]
> > Also I had to introduce a field "Interface Devices" into the network overview, because there was no way user could see what interface devices they just assigned during network creation. Link to various network devices are clickable and will redirect you to networks device's page
> > Screenshot from 2020-03-10 13-45-32
> 
> FWIW, the tab spacing is weird here, and the delete button should be on the inside, I think. As it is right now, on Cockpit, we have the default action on the edge (in this case, the right side), so it should be Delete on the left and Activate/Deactivate on the right.
> 
> It's even weird to have a tab at all, when there's only one tab. (I know that's a limitation of our current list implementation. There's a work-around of hiding the tab with CSS though, I guess...?)
> 
> However, we're planning to change it (in at least some places) soon to have the default on the row itself with a … menu for other stuff like delete. But that's near-future stuff, not current stuff. And I don't want it to get in the way of this PR.
> 
> Doing this also would mean that we wouldn't mix top-level actions with the tab line, which should make things clearer. But we should probably introduce the design pattern in multiple places at the same, within the same release, if we can.
> 
> TL;DR: Swap button order. Figure out the reason why the tab is floating. (I can probably do the latter one... but I think it's because the buttons are full size, not our smaller versions of the buttons. If we simply hide the tab in this case, that's another way to "fix" it. Probably.)

Here is how it looks now:

Screenshot from 2021-02-15 14-49-42

**Comment by garrett** [comment]
I can't select the items by clicking on them. I have to specifically click the checkboxes. It's weird (and wrong).

Perhaps it's related to the labels having a `for` attribute and also wrapping the input? When I remove the `for` in the DOM, it works. It's supposed to be one or the other, it doesn't need to be both.

Aha, I see the for and the id are different:

- Label's `for`: `create-network-bridge-iface-redhat0`
- Input's `id`: `create-network-bridge-iface-redhat0-checkbox`

---

Also, the checkboxes are all placed wrongly in Firefox. I think this might be a PatternFly bug, where they're assuming all checkboxes are the same size as Chrome on a Mac. I have to do additional browser-testing, but we might need to set the size manually. I'm considering this for a work-around in Cockpit (`pkg/lib/patternfly/patternfly-4-overrides.scss`) and opening an issue for PF if so:

 ```css
// PF assumes check sizes are Chrome-based; size needs to be set for cross-browser compatibility
.pf-c-check > .pf-c-check__input {
    height: var(--pf-c-check__label--FontSize);
}
```

But it's late here, so I need to stop for today.

If I'm right about the label sizing, it should be a separate commit (and probably even a separate PR). It does impact this one though, hence mentioning it here.

**Comment by skobyda** [comment]
Nice catch. The `id` and `for` fixed the clickable row and yes PF4 override works. I added it as a separate commit, but feel free to create a PR jointed to a PF4 issue.
Screenshot from 2021-02-16 01-56-29

File: pkg/lib/patternfly/patternfly-4-overrides.scss [hunk 2]
```diff
@@ -76,3 +76,8 @@ svg {
       }
   }
 }
+
+// PF assumes check sizes are Chrome-based; size needs to be set for cross-browser compatibility
```
Comment (KKoukiou) [inline pkg/lib/patternfly/patternfly-4-overrides.scss:80]: Right as you mentioned already, @garrett this will need an upstream issue and the l link here in the code :)

Comment (garrett) [reply pkg/lib/patternfly/patternfly-4-overrides.scss:80]: Yep, currently working on making sure there's a solution that works across browsers and can be upstreamed to PF. What's in this PR isn't that (yet).

I have something that seems to work correctly across all browsers now; I just need to convert it back to PF variables and ensure that works too.

**Comment by garrett** [comment]
Here's an updated version that fixes the math behind the sizing and placement of the widget in a cross-browser way:

```scss
// PF assumes check sizes are Chrome-based; size needs to be set for cross-browser compatibility
.pf-c-check > .pf-c-check__input {
    // Set the height of the input widget to be the same as its label
    --pf-c-check__input--Height: var(--pf-c-check__label--FontSize);
    // Offset by the top half of the difference between the height and the line height
    --pf-c-check__input--MarginTop: calc((var(--pf-c-check__label--LineHeight) - 1) / 2 * var(--pf-c-check__input--Height));
}
```

Sorry it took a while; I needed to figure out what values worked everywhere and why, then figure out which PF variables to use and which to adjust. I think this is it? And I've included the reasoning as comments in the SCSS.

I'll adjust the commit in this PR to this new version.

So this fixes the input alignment for nested checkboxes; the same logic should probably also apply to radio buttons — and then what about non-nested versions? :thinking:

**Comment by skobyda** [comment]
Thank you for your fix @garrett (When I cherry picked it to my local branch because I wanted to fix tests, the commit lost your authorship, weird. But it's still the same)
@KKoukiou I think this is ready for another re-review

**Comment by garrett** [comment]
@skobyda: I reworked 1cfc5af and moved it to its own PR @ https://github.com/cockpit-project/cockpit/pull/15363 — it now has both check and radio support and supports chimera pages (those pages that still mix both PF3 + PF4). So we could safely drop the commit from this PR.

Hopefully we'll land the other PR soon.

**Comment by garrett** [comment]
(Oh, by that last comment, I mean we can drop the commit here and rely on that other one for the fix. Just to clarify things. Thanks!)

**Comment by skobyda** [comment]
Dropped the commit.
After #15363 gets merged, this PR is ready for re-review

**Review by garrett — APPROVED** [review]
LGTM! :+1:

**Comment by martinpitt** [comment]
@KKoukiou do you have some time to review this? It seems this is now ready. However, this is now over two weeks old -- @skobyda , can you please rebase to current master to get an up to date test run?

**Comment by KKoukiou** [comment]
@skobyda you need to move this to the new repository. Please reopen there.

(sorry I just noticed this was just missing my review)
